File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int[][] mirro;
    static boolean[][] visited;
    static int[] dirY = {1, 0, -1, 0};
    static int[] dirX = {0, 1, 0, -1};
    static int h;
    static int w;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        h = Integer.parseInt(st.nextToken());
        w = Integer.parseInt(st.nextToken());
        mirro = new int[h+1][w+1];
        visited = new boolean[h+1][w+1];

        for(int y=1; y<=h; y++){
            String line = br.readLine();
            for(int x=1; x<=w; x++){
                mirro[y][x] = Integer.parseInt(line.substring(x-1, x));
            }
        }

        bfs();
        System.out.println(mirro[h][w]);
    }

    public static void bfs(){
        Queue<Integer[]> que = new LinkedList<>();
        que.add(new Integer[]{1,1});
        visited[1][1] = true;

        while(!que.isEmpty()){
            int y = que.peek()[0];
            int x = que.peek()[1];
            que.poll();
            int count = mirro[y][x];

            for(int i=0; i<4; i++){
                int newY = y+dirY[i];
                int newX = x+dirX[i];

                if(newY < 0 || newY > h || newX < 0 || newX > w || visited[newY][newX]) continue;
                if(mirro[newY][newX] == 1){
                    que.add(new Integer[]{newY, newX});
                    visited[newY][newX] = true;
                    mirro[newY][newX] = count + 1;
                }
            }
        }
    }
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input.shift().split(' ').map(Number);
const map = [];
const visited = Array.from({length: n}, () => Array(m).fill(false));
const dirY = [-1, 0, +1, 0];
const dirX = [0, +1, 0, -1];

for (let i = 0; i < n; i++) {
    map[i] = input[i].split('').map(Number);
}

function bfs() {
    const que = [];
    que.push([0, 0, 1]);
    visited[0][0] = true;

    while (que.length) {
        const [y, x, move] = que.shift();
        if(y === n-1 && x === m-1) return move;

        for (let i = 0; i < 4; i++) {
            let newY = y + dirY[i];
            let newX = x + dirX[i];

            if(newY < 0 || newY >= n || newX < 0 || newX >= m) continue;
            if(visited[newY][newX] || map[newY][newX] === 0) continue;

            que.push([newY, newX, move+1]);
            visited[newY][newX] = true;
        }
    }
}

console.log(bfs());